tests/jetstream_test.ts

```diff
@@ -3010,3 +3010,99 @@ Deno.test("jetstream - pull next", async () => {
 
   await cleanup(ns, nc);
 });
+
+Deno.test("jetstream - pull errors", async () => {
+  const { ns, nc } = await setup(jetstreamServerConf({}, true));
+
+  const { stream, subj } = await initStream(nc);
+  const jsm = await nc.jetstreamManager();
+
+  await jsm.consumers.add(stream, {
+    durable_name: "me",
+    ack_policy: AckPolicy.Explicit,
+  });
+  const js = nc.jetstream();
+
+  async function expectError(
+    expires: number,
+    code: ErrorCode,
+  ) {
+    try {
+      await js.pull(stream, "me", expires);
+    } catch (err) {
+      assertEquals(err.code, code);
+    }
+  }
+
+  await expectError(0, ErrorCode.JetStream404NoMessages);
+  await expectError(1000, ErrorCode.JetStream408RequestTimeout);
+
+  await js.publish(subj);
+
+  // we expect a message
+  const a = await js.pull(stream, "me", 1000);
+  assertEquals(a.seq, 1);
+
+  await cleanup(ns, nc);
+});
+
+Deno.test("jetstream - pull error: max_waiting", async () => {
+  const { ns, nc } = await setup(jetstreamServerConf({}, true));
+  if (await notCompatible(ns, nc, "2.8.2")) {
+    return;
+  }
+
+  const { stream } = await initStream(nc);
+  const jsm = await nc.jetstreamManager();
+
+  await jsm.consumers.add(stream, {
+    durable_name: "me",
+    ack_policy: AckPolicy.Explicit,
+    max_waiting: 1,
+  });
+  const js = nc.jetstream();
+
+  async function expectError(
+    expires: number,
+    code: ErrorCode,
+    message?: string,
+  ) {
+    try {
+      await js.pull(stream, "me", expires);
+    } catch (err) {
+      assertEquals(err.code, code);
+      if (message) {
+        assertEquals(err.message, message);
+      }
+    }
+  }
+  await Promise.all([
+    expectError(
+      3000,
+      ErrorCode.JetStream408RequestTimeout,
+      "408 request canceled",
```

Comment:
So as of server rev (https://github.com/nats-io/nats-server/commit/cadf921ed162951e69b6a031648b107c83f63241), when you reach the maximum number of pull requests, the new request will get a 409 (Exceeded MaxWaiting), and not an old one getting a 408 (Request Canceled). This was incorrect behavior. The server release notes for the upcoming 2.8.2 will state:
```
### Changed
- JetStream:
    - When reaching the max number of pull requests, the server would
      evict existing pending requests (with a `408 Request Canceled`) 
      instead of sending a `409 Exceeded MaxWaiting` to the new request (#3099)
```